### PR TITLE
Add golang scripts to work with GPG signatures

### PR DESCRIPTION
**Pull request by tomasmik**

### CODE REVIEW

**Comment by tomasmik** [comment]
See the same scripts in action in my repository:
* [Actions](https://github.com/tomasmik-testorg/testing-signatures/actions)
* [Releases](https://github.com/tomasmik-testorg/testing-signatures/releases)

Used both of these to generate a new key and use it to sign a release:
image

**Review by Yantrio — APPROVED** [review]

File: .github/workflows/sign.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    - name: Import GPG Private Key
+      run: |
+        echo "${{ secrets.GPG_PRIVATE_KEY }}" | base64 --decode | gpg --import
+      env:
+        GPG_TTY: /dev/ttys000  # Set the GPG_TTY to avoid issues with pinentry
+
+    - name: Build and run Go script
+      run: |
+        OWNER="${{ github.event.inputs.owner}}"
```
Comment (Yantrio) [inline .github/workflows/sign.yml:40]: tiny nit, whitepsace is a bit strange here

File: .github/workflows/sign.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
+
```
Comment (cube2222) [inline .github/workflows/sign.yml:1]: I think there's an unnecessary newline here.

File: go/gpg/main.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+	for _, id := range es.Identities {
+		id.SelfSignature = &packet.Signature{
+			CreationTime:              time.Now(),
+			KeyLifetimeSecs:           &keyLifetime,
+			SigType:                   packet.SigTypePositiveCert,
+			PubKeyAlgo:                packet.PubKeyAlgoRSA,
+			Hash:                      config.Hash(),
+			IssuerKeyId:               &es.PrimaryKey.KeyId,
+			PreferredHash:             []uint8{8},
```
Comment (cube2222) [inline go/gpg/main.go:65]: I think a comment on these anonymous byte arrays would be good

File: go/gpg/main.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		Name:           *secretName,
+		KeyID:          *k.KeyID,
+		EncryptedValue: en,
+		Visibility:     *visibility,
+	})
+
+	return err
+}
+
+func encodeWithPublicKey(text string, publicKey string) (string, error) {
```
Comment (cube2222) [inline go/gpg/main.go:117]: I don't understand. Why are we encrypting the newly generated private gpg key with some random public key? Is this some GitHub Secrets API-specific thing?

Comment (cube2222) [reply go/gpg/main.go:117]: Anyhow, could use a comment

File: go/gpg/main.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+		CompressionConfig: &packet.CompressionConfig{
+			Level: packet.BestCompression,
+		},
+		RSABits: 4096,
+	}
+
+	es, err := openpgp.NewEntity(*gpgName, *gpgComment, *gpgEmail, config)
+	if err != nil {
+		fmt.Println(err)
+		return
```
Comment (cube2222) [inline go/gpg/main.go:52]: Why not just fatal like in other places?

File: go/gpg/main.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	privateKeyBuffer := new(bytes.Buffer)
+	err = es.SerializePrivate(privateKeyBuffer, config)
+	if err != nil {
+		log.Fatalf("Error serializing key: %v", err)
+	}
+
+	err = setSecret(context.Background(), privateKeyBuffer.Bytes())
+	if err != nil {
+		log.Fatalf("Error uploading GPG key to GitHub: %v", err)
+	}
```
Comment (cube2222) [inline go/gpg/main.go:90]: It seems like we're never exporting the public key that users need to verify the signatures, no?

I think this should print the public key and/or fingerprint.

File: go/gpg/main.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	es, err := openpgp.NewEntity(*gpgName, *gpgComment, *gpgEmail, config)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	var keyLifetime uint32 = 3 * 365 * 24 * 60 * 60
+
+	for _, id := range es.Identities {
+		id.SelfSignature = &packet.Signature{
```
Comment (cube2222) [inline go/gpg/main.go:58]: What are those self-signatures, could you please explain in a comment?

File: go/gpg/main.go (hunk 4 above)
Comment (tomasmik) [reply go/gpg/main.go:117]: >Is this some GitHub Secrets API-specific thing?

You're totally right here. This is github specific. It's basically how you store a secret.

File: go/gpg/main.go (hunk 5 above)
Comment (tomasmik) [reply go/gpg/main.go:52]: My bad! Should totally be just that.

File: go/gpg/main.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+			CreationTime:              time.Now(),
+			KeyLifetimeSecs:           &keyLifetime,
+			SigType:                   packet.SigTypePositiveCert,
+			PubKeyAlgo:                packet.PubKeyAlgoRSA,
+			Hash:                      config.Hash(),
+			IssuerKeyId:               &es.PrimaryKey.KeyId,
+			PreferredHash:             []uint8{8},
+			PreferredCompression:      []uint8{2},
+			PreferredSymmetric:        []uint8{9},
```
Comment (Yantrio) [inline go/gpg/main.go:67]:
```suggestion
			SigType:                   packet.SigTypePositiveCert, // 0x13 [https://datatracker.ietf.org/doc/html/rfc4880#section-5.2.1]
			PubKeyAlgo:                packet.PubKeyAlgoRSA,  // 1 for RSA (Encrypt or Sign) [https://datatracker.ietf.org/doc/html/rfc4880#section-9.1]
			Hash:                      config.Hash(), 
			IssuerKeyId:               &es.PrimaryKey.KeyId,
			PreferredHash:             []uint8{8}, // 8 for SHA256 (Encrypt or Sign) [https://datatracker.ietf.org/doc/html/rfc4880#section-9.4]
			PreferredCompression:      []uint8{2}, // 2 for ZLIB [https://datatracker.ietf.org/doc/html/rfc4880#section-9.3]
			PreferredSymmetric:        []uint8{9}, // 9 for AES with 256-bit key [https://datatracker.ietf.org/doc/html/rfc4880#section-9.2]
```


This should be the comments that @cube2222 requested in [this](https://github.com/opentofu/scripts/pull/3/files#r1389516553) review comment

File: go/env/main.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+	pat = os.Getenv("GITHUB_PAT")
+	if pat == "" {
+		panic("GITHUB_PAT environment variable not set")
+	}
+
+	secretValue = os.Getenv("SECRET_VALUE")
+	if secretValue == "" {
+		panic("SECRET_VALUE environment variable not set")
+	}
+
```
Comment (cube2222) [inline go/env/main.go:41]: Let's also please check that repo is not an empty string.

File: go/gpg/main.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+	flag.Parse()
+
+	config := &packet.Config{
+		DefaultCipher:          packet.CipherAES256,    // 9 for AES with 256-bit key [https://datatracker.ietf.org/doc/html/rfc4880#section-9.2]
+		DefaultCompressionAlgo: packet.CompressionZLIB, // 2 for ZLIB [https://datatracker.ietf.org/doc/html/rfc4880#section-9.3]
+		CompressionConfig: &packet.CompressionConfig{
+			Level: packet.BestCompression, // 9 for best compression [https://datatracker.ietf.org/doc/html/rfc4880#section-9.3]
+		},
+		RSABits: 4096, // 4096-bit RSA key [https://datatracker.ietf.org/doc/html/rfc4880#section-
+	}
```
Comment (cube2222) [inline go/gpg/main.go:47]: Now that you use proper const, you don't need the `9 for AES with 256-bit key` parts.

But the links I'd propose keeping, those are very useful.

File: go/gpg/main.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+		//
+		// After setting the SelfSignature, the user ID (name and email) is signed with the entity's
+		// primary key and private key, creating a verifiable signature for future reference.
+		//
+		// This process is crucial in OpenPGP as it allows the owner of an identity to claim
+		// ownership of a public key, and for others to verify that claim.
+		id.SelfSignature = &packet.Signature{
+			CreationTime:              time.Now(),
+			KeyLifetimeSecs:           &keyLifetime,
+			SigType:                   packet.SigTypePositiveCert, // 0x13 [https://datatracker.ietf.org/doc/html/rfc4880#section-5.2.1]
```
Comment (cube2222) [inline go/gpg/main.go:72]: Same regarding consts where they're not literals, but please keep links.

File: go/gpg/main.go [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+			PubKeyAlgo:                packet.PubKeyAlgoRSA,       // 1 for RSA (Encrypt or Sign) [https://datatracker.ietf.org/doc/html/rfc4880#section-9.1]
+			Hash:                      config.Hash(),
+			IssuerKeyId:               &es.PrimaryKey.KeyId, // Key ID of the signing key
+			PreferredHash:             []uint8{8},           // 8 for SHA256 (Encrypt or Sign) [https://datatracker.ietf.org/doc/html/rfc4880#section-9.4]
+			PreferredCompression:      []uint8{2},           // 2 for ZLIB [https://datatracker.ietf.org/doc/html/rfc4880#section-9.3]
```
Comment (cube2222) [inline go/gpg/main.go:77]: Can't we use the same const we use above here? That is, packet.CompressionZLIB

**Pull request merged**